### ISSUE TITLE
Wrong result when performing variance calculation

### ISSUE THREAD

**Issue by BeStrongok** [secretflow/secretflow#155]
### Issue Type

Others

### Source

binary

### Secretflow Version

secretflow 0.7.7b1

### OS Platform and Distribution

Linux ubuntu 18.04

### Python version

3.8.13

### What happend and What you expected to happen.

```shell
Hi,
I got the wrong result when performing Variance calculation. Should i modify the code or add some configurations?
```
The code is:
```
spu_device = sf.SPU(cdef)


alice, bob, carol = sf.PYU('alice'), sf.PYU('bob'), sf.PYU('carol')


v_alice_path = "/shared/data0.csv"
v_bob_path = "/shared/sf_test/data1.csv"
v_carol_path = "/shared/data2.csv"

import jax.numpy as np

def load_data(path):
    x = pd.read_csv(path, index_col=0)
    return x.to_numpy()

def mul(a, b, c):
    tmp = np.concatenate((a, b, c), axis=0)
    return np.var(tmp)

alice_spu = alice(load_data)(v_alice_path).to(spu_device)
bob_spu = bob(load_data)(v_bob_path).to(spu_device)
carol_spu = carol(load_data)(v_carol_path).to(spu_device)

start = time.time()
res = spu_device(mul)(alice_spu, bob_spu, carol_spu)
res_public = sf.reveal(res)
print(f"mpc time: {time.time() - start}")

print(res_public)
```
The results i got were:
```
2022-10-17 12:00:56.141604: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
mpc time: 15.221762895584106
0.0
```

### Reproduction code to reproduce the issue.

```shell
The configuration of SPU is:
    'runtime_config': {
        'protocol': spu.spu_pb2.ABY3,
        'field': spu.spu_pb2.FM64,
        'fxp_fraction_bits': 18
    }
```

**Comment by anakinxc** [comment]
Hi,

Can you provide a tiny repro example?

Thanks

**Comment by BeStrongok** [comment]
> Hi,
> 
> Can you provide a tiny repro example?
> 
> Thanks

Yes, i generat data with the code:
```
import sys
import numpy as np
import pandas as pd

np.random.seed(888)
test_idx = sys.argv[1]
player_index = int(sys.argv[2])

sum_all = []
data_num = {'10t': 10000, '10m': 10000000, '100m': 100000000}[test_idx]

data = {"data{}".format(player_index): np.random.rand(data_num)}
data_df = pd.DataFrame(data)
data_df.to_csv("data{}.csv".format(player_index))
```

**Comment by anakinxc** [comment]
Cool beans!

I'll take a look.

Best

**Comment by anakinxc** [comment]
Hi BeStrongok,

I have confirmed this is a numeric issue.

We will try to systemically improve this next release. 

For now, you can try to increase fxp_fraction_bits to improve accuracy.

Thanks

**Comment by BeStrongok** [comment]
> Hi BeStrongok,
> 
> I have confirmed this is a numeric issue.
> 
> We will try to systemically improve this next release.
> 
> For now, you can try to increase fxp_fraction_bits to improve accuracy.
> 
> Thanks

Thanks, i increased the fxp_fraction_bits to 32 and the result of variance is `0.002210606`, which was not correct.
Different fxp_fraction_bits can cause very different results.

**Comment by anakinxc** [comment]
> > Hi BeStrongok,
> > I have confirmed this is a numeric issue.
> > We will try to systemically improve this next release.
> > For now, you can try to increase fxp_fraction_bits to improve accuracy.
> > Thanks
> 
> Thanks, i increased the fxp_fraction_bits to 32 and the result of variance is `0.002210606`, which was not correct. Different fxp_fraction_bits can cause very different results.

Increase fxp_fraction_bits can only improve accuracy to a certain point, can you also try with  FM128 (which has default fxp_fraction_bits = 26)?

**Comment by BeStrongok** [comment]
> 

Thank you very much, i got the correct result with `FM128 ` and `fxp_fraction_bits=32`. :)

**Comment by anakinxc** [comment]
> > 
> 
> Thank you very much, i got the correct result with `FM128 ` and `fxp_fraction_bits=32`. :)

Glad to hear that. We will keep tracking this internally.
